File: benchmarks/fixtures/webapp_rb/services/auth_service.rb
# High-level authentication service.

require_relative '../utils/helpers'
require_relative '../auth/service'
require_relative '../auth/tokens'
require_relative '../database/connection'
require_relative '../database/queries'
require_relative '../events/dispatcher'
require_relative '../exceptions'

# Orchestrates authentication flows.
class AuthenticationService < BaseService
  def initialize(db, events)
    super()
    @auth = AuthService.new(db)
    @users = UserQueries.new(db)
    @sessions = SessionQueries.new(db)
    @events = events
  end

  # Authenticate a user — main entry point for login flow.
  def authenticate(email, password, ip = 'unknown')
    _log("Authentication attempt for #{email}")
    clean_email = sanitize_input(email)
    raise ValidationError.new('Email is required', field: :email) if clean_email.empty?

    begin
      token = @auth.login(clean_email, password)
      unless token
        @events.emit('auth.login_failed', { email: clean_email, ip: ip })
        raise AuthenticationError.new('Invalid credentials')
      end
      @events.emit('auth.login_success', { email: clean_email, ip: ip })
      { token: token, email: clean_email }
    rescue AuthenticationError
      raise
    rescue StandardError => e
      @events.emit('auth.login_failed', { email: clean_email, ip: ip })
      raise AuthenticationError.new("Authentication failed: #{e}")
    end
  end

  # Verify a token.
  def verify_token(token)
    begin
      user = validate_token(token)
      @users.find_by_email(user.email) if user
    rescue TokenError
      nil
    end
  end

  # Log out.
  def do_logout(token)
    _log('Processing logout')
    session = @sessions.find_active_session(token)
    if session
      @sessions.expire_session(session[:id])
      @events.emit('auth.logout', { session_id: session[:id] })
      return true
    end
    false
  end
end

File: benchmarks/fixtures/webapp_php/services/auth_service.php
<?php

namespace App\Services;

use App\AuthenticationError;
use App\Auth\AuthService;
use App\Auth\BaseService;
use App\Auth\TokenError;
use App\Database\DatabaseConnection;
use App\Database\SessionQueries;
use App\Database\UserQueries;
use App\Events\EventDispatcher;
use App\ValidationError;

use function App\Auth\validate_token;
use function App\Utils\sanitize_input;

/**
 * Orchestrates authentication flows for the services layer.
 */
class AuthenticationService extends BaseService
{
    private AuthService $auth;
    private UserQueries $users;
    private SessionQueries $sessions;

    public function __construct(DatabaseConnection $db, private EventDispatcher $events)
    {
        parent::__construct();
        $this->auth = new AuthService($db);
        $this->users = new UserQueries($db);
        $this->sessions = new SessionQueries($db);
    }

    /**
     * Authenticate a user — main entry point for the login flow.
     *
     * @return array<string, mixed>
     */
    public function authenticate(string $email, string $password, string $ip = 'unknown'): array
    {
        $this->log("Authentication attempt for {$email}");
        $cleanEmail = sanitize_input($email);
        if ($cleanEmail === '') {
            throw new ValidationError('Email is required', 'email');
        }

        $token = $this->auth->login($cleanEmail, $password);
        if ($token === null) {
            $this->events->emit('auth.login_failed', ['email' => $cleanEmail, 'ip' => $ip]);
            throw new AuthenticationError('Invalid credentials');
        }
        $this->events->emit('auth.login_success', ['email' => $cleanEmail, 'ip' => $ip]);

        return ['token' => $token, 'email' => $cleanEmail];
    }

    /**
     * Verify a token and return the matching user record.
     *
     * @return array<string, mixed>|null
     */
    public function verifyToken(string $token): ?array
    {
        try {
            $user = validate_token($token);

            return $this->users->findByEmail($user->email);
        } catch (TokenError) {
            return null;
        }
    }

    public function doLogout(string $token): bool
    {
        $this->log('Processing logout');
        $session = $this->sessions->findActiveSession($token);
        if ($session !== null) {
            $this->sessions->expireSession((int) $session['id']);
            $this->events->emit('auth.logout', ['session_id' => $session['id']]);

            return true;
        }

        return false;
    }
}
